Implemented appointment sheet #3785

stgallagher opened the pull request:
### Summary

Appointment sheet that lists a variety of useful information for administering an event.
##### Actions
- Linked the appointment sheet to an icon next to a participant's name in the scheduled activities tab of the welcome page.
- Added a controller, model and view for the appointment sheet
- Placed as much of the logic to derive the information necessary for the sheet in the model. There are a few variables(Event time, Event date, Data Collector name) that I had to do straight from the controller since they are derived from the PSC scheduled activities report. I am looking into ways of migrating these to the model.
##### Comments
- There are some fields in the appointment sheet design mock-up that I couldn't get given that their source objects don't exist prior to an event's execution. Examples are Contact Type, Contact Comments. 
- There are some other fields where the source data doesn't exist or would require a fair amount of work to create, e.g. Pregnancy Length, Communication Preferences, Visit Coordinator name.
#### Design Mock-up

![Screen Shot 2013-04-10 at 8 58 37 AM](https://f.cloud.github.com/assets/801580/374669/252e2c54-a3a4-11e2-89dc-7d526176cf9d.png)
#### Implementation Screenshot

![Screen Shot 2013-04-12 at 2 06 06 PM](https://f.cloud.github.com/assets/801580/374671/32c882b0-a3a4-11e2-87c7-2e2782ece86b.png)
#### The printed PDF version (in Firefox) (Admittedly needs work)

![Screen Shot 2013-04-12 at 2 24 31 PM](https://f.cloud.github.com/assets/801580/374738/befd1d70-a3a6-11e2-89d0-81bc3fa61023.png)

stgallagher committed:
implemented appointment sheet #3785

stgallagher committed:
make specs run 3x faster #3785

rsutphin commented:
The implementation screenshot shows a person ID (XXXX-XXXX-XXXX), not a participant ID (XXX-XX-XXXX) under the participant's name.

hannahwhy commented on app/assets/stylesheets/partials/_appointment_sheet.scss:
```diff
@@ -0,0 +1,70 @@
```
These values should be derived from existing Cases metrics, not duplicated here.

rsutphin commented on app/models/event.rb:
```diff
@@ -323,14 +322,6 @@ def strip_time_whitespace
-    end
-  end
-  private :set_start_time
-
```
Why was this removed? It seems like a change with far greater impact than adding a new view implies.

stgallagher replied:
It was before I knew that event_start_time != scheduled event time. I talked to Paul and he said it was an out-dated means of setting event_start_time and that they are set by PSC now. I don't need this change anymore since I don't use event_start_time. I will reverse the removal.

hannahwhy commented on app/assets/stylesheets/partials/_appointment_sheet.scss:
```diff
@@ -0,0 +1,70 @@
+  font-size: 130%;
+}
+
+.column-container {
```
At least based on this, you're using the `@column` mixin in something that isn't a container.  That might work, but it isn't intended usage.

rsutphin commented on app/models/appointment_sheet.rb:
```diff
@@ -0,0 +1,146 @@
+#  updated_at :datetime
+#
+
+class AppointmentSheet < ActiveRecord::Base
```
You can have a model that isn't an AR model. Since there's nothing persisted in this feature (AFAICT), you don't need to descend from AR::Base.

stgallagher replied:
Understood. Will change.

hannahwhy commented on app/assets/stylesheets/partials/_appointment_sheet.scss:
```diff
@@ -0,0 +1,70 @@
+
+}
+
+.large-and-bolded {
```
This class name describes a visual effect; it doesn't describe what the class members _are_.  If you're using this class on something that is supposed to be a heading, use the appropriate heading element and develop a style with an appropriate selector.

hannahwhy commented on app/assets/stylesheets/partials/_appointment_sheet.scss:
```diff
@@ -0,0 +1,70 @@
+  font-weight: bold;
+}
+
+.column-one {
```
Same comments as those on `large-and-bolded`.

hannahwhy commented on app/assets/stylesheets/partials/_appointment_sheet.scss:
```diff
@@ -0,0 +1,70 @@
+  margin-left: 5em;
+}
+
+.column-two {
```
Same comments as those on `large-and-bolded`.

hannahwhy commented on app/controllers/welcome_controller.rb:
```diff
@@ -122,8 +122,8 @@ def destroy_participant_and_redirect(participant, resp, destroy_person = true)
     def get_scheduled_activities_report(options = {})
-      @start_date = 1.day.ago.to_date.to_s
-      @end_date   = params[:end_date] || 6.weeks.from_now.to_date.to_s
+      @start_date = Date.new(2013,4,8)#Date.today.beginning_of_week
```
Why is this date hardcoded?

stgallagher replied:
The appointment sheet comes from the scheduled activities page which (according to the design mock-up) is scaled for whatever current week, which changes dynamically (its the commented out code) but I had a series of test subjects assigned to dates so, to keep from having to update their dates, I hard-coded these dates. I do need to change this before merging.

hannahwhy commented on app/models/appointment_sheet.rb:
```diff
@@ -0,0 +1,146 @@
+      consents << participant_consents(child)
+    end
+    consents
+  end
```
``` ruby
def child_consents
  @person.participant.children.each_with_object([]) { |c, a| a << participant_consents(c) }
end
```

rsutphin replied:
Or even:

``` ruby
def child_consents
  @person.participant.children.collect { |c| participant_consents(c) }
end
```

hannahwhy replied:
Looking at this as a mapping, I'm now wondering if the result ought to be flattened, or if it's intentionally an array of arrays.

stgallagher replied:
There could be several children and they can each have several consents, so  I think conceptually a two-dimensional array is a good fit.

hannahwhy commented on app/models/appointment_sheet.rb:
```diff
@@ -0,0 +1,146 @@
+        when 0 then nil
+        when 1 then "#{value} #{unit}"
+        else "#{value} #{unit}s"
+      end
```
ActiveSupport's inflector supplies `String#pluralize`; why not use that?

stgallagher replied:
Good idea. Thanks.

hannahwhy commented on app/models/appointment_sheet.rb:
```diff
@@ -0,0 +1,146 @@
+      ages << age(child.person_dob_date)
+    end
+    ages
+  end
```
https://github.com/NUBIC/ncs_navigator_core/pull/127/files#r3778970

hannahwhy commented on app/helpers/welcome_helper.rb:
```diff
@@ -16,10 +16,15 @@ def activities(json_response)
+    empties = activities.select { |act| act.activity_time.nil?  }
+
+    sorted + empties
   end
```
``` ruby
def sort(activities)
  activities.sort_by { |a| a.activity_time.blank? ? Float::INFINITY : a.activity_time.delete(':').to_i }
end
```

hannahwhy commented on app/views/appointment_sheets/show.html.haml:
```diff
@@ -0,0 +1,66 @@
+.column-container
+  .column-one
+    .large-and-bolded
+      = "#{@sheet.event_type}"
```
String interpolation isn't required here.

stgallagher replied:
Good point. Will change.

rsutphin commented on app/helpers/welcome_helper.rb:
```diff
@@ -16,10 +16,15 @@ def activities(json_response)
-                               a1.send(:activity_time) <=> a2.send(:activity_time) :
-                             ( a1.send(:activity_time) ? -1 : 1 ) }
+  def sort_activities_chronologically_with_blank_times_at_the_bottom_of_list(activities)
+    convert_time_to_integer = Proc.new { |time| time.delete(":").to_i }
```
These are zero-padded 24-hour time strings, right? You shouldn't need to convert them to integers to sort them.

stgallagher replied:
They are _not_ zero-padded, which created the need to do extra stuff for their sorting.

stgallagher committed:
cope with missing info #3785

Give a nice message if info is missing. Change 
some of the info fetchers to return nil if their
source objects don't exist

stgallagher committed:
removed controller test #3785

In lieu of integration test

stgallagher committed:
made css more semantic #3785

Deleted inherited column classes. Made proportions
and page element flow more reliant on local 
classes and ids.

stgallagher committed:
made date range configurable #3785

Also reverted the scheduled activities tab to its
default 6 week range if parameters are not 
included.

stgallagher committed:
adapted methods for nil #3785

Not a persisted model, so removed inheritance 
from ActiveRecord::Base. Refactored #age to use
#pluralize

stgallagher committed:
restored #set_start_time and hook #3785

stgallagher committed:
moved "not found" logic to the view #3785

The display of "<whatever> is not found" is a view 
concern so I put most of the logic for that 
function in the view code, relying on the models
to return nil when records aren't found.

stgallagher committed:
title descriptor for appointment sheet links #3785

Also I needed a target for capybara.

stgallagher committed:
revised specs for appointment sheet #3785

stgallagher committed:
mocking out calls to PSC #3785

Needed to cover #scheduled_activities and 
#build_activity_plan

stgallagher committed:
features for appointment sheet #3785

With step definitions.

stgallagher committed:
Updated selenium-webdriver gem #3785

When using the @javascript tag on a scenario, the
automated browser window would open and then 
freeze. Updating the gem fixed the issue.

stgallagher commented:
Ready for Review.

### Summary

Responded to the issues brought up in the first round of code review. Added Cucumber integration tests. 

##### Actions
- Changed erroneous `person_id` to `participant_id`
- Restored `#set_start_time` to `Event` after its unnecessary deletion
- Removed improper CSS inheritence
- Made CSS selectors more semantic
- Date range for scheduled activities now configurable through URL parameters
- Restored default date range to 6 weeks
- Refactored `AppointmentSheet#participant_consents` and `AppointmentSheet#age`
- Adapted methods to handle `nil` conditions more appropriately
- Added integration tests for viewing upcoming appointments and viewing appointment sheets

rsutphin commented on spec/models/address_spec.rb:
```diff
@@ -53,17 +53,18 @@
 
   it "should describe itself" do
-    addr.to_s.should == "#{addr.state}"
+    state_name = "#{addr.state}".downcase.capitalize
```
Expectations should be literal and concrete as much as possible. This just recapitulates what's in the code.

rsutphin commented on app/models/address.rb:
```diff
@@ -98,7 +98,7 @@ def to_s(separator = ' ')
     addr << unit
     city_state << city
-    city_state << state.to_s if state && state.local_code != -4
+    city_state << state.to_s.downcase.capitalize if state && state.local_code != -4
```
Missed this the first time around: there exist states which have more than one word in their names. You probably want `titleize` here. Note also that `capitalize` and `titleize` both take care of changing the case for all the letters in the string. No need to `downcase` first.

rsutphin commented on db/schema.rb:
```diff
@@ -79,6 +79,11 @@
+    t.datetime "created_at"
+    t.datetime "updated_at"
+  end
+
```
This is still here.

stgallagher replied:
Forgot that one. Deleting it and the migration file. Thanks.

rsutphin commented on db/migrate/20130409233256_create_appointment_sheets.rb:
```diff
@@ -0,0 +1,8 @@
+      t.timestamps
+    end
+  end
+end
```
This is still here.

hannahwhy commented on app/assets/stylesheets/partials/_appointment_sheet.scss:
```diff
@@ -0,0 +1,81 @@
+
+#children {
+  margin-top: 1em;
+}
```
Common style motifs (and this is one) should be factored out into SCSS mixins.  The mixin name, like the style name, should reflect the meaning of the elements to which is applies.

http://sass-lang.com/docs/yardoc/file.SASS_REFERENCE.html#mixins

hannahwhy commented on features/appointment_sheet/viewing_an_appointment_sheet.feature:
```diff
@@ -0,0 +1,75 @@
+    Then I should see the the "Birth Visit Information Sheet" as to be conducted
+    Then I should see the the "Pregnancy Visit 1 Information Sheet" as to be conducted
+    Then I should see the the "Pregnancy Health Care Log" as to be conducted
+    Then I should the next event as "Pregnancy Visit 2"
```
This is the sort of thing that tables are very good at.

hannahwhy commented on features/appointment_sheet/viewing_upcoming_appointments.feature:
```diff
@@ -0,0 +1,50 @@
+    Given an authenticated user
+    When I view my scheduled events for 2013-04-08 to 2013-04-13
+    Then I should see "Dawn Davidson" scheduled for "Monday April 8" for a "Pregnancy Visit 1" event at "2:45 PM"
+    Then I should see "Elaine Edwinson" scheduled for "Wednesday April 10" for a "Pbs Participant Eligibility Screening" event at "9:30 AM"
```
When chaining `Then`s (if you must -- I prefer that each scenario have a single `Then`, though sometimes it's neither convenient nor readable) use `And`.  The same applies for Given and When.

Keep in mind that a scenario is meant to be read _by a human_.  (So is source code, for that matter.)  Ideally, that human is a domain expert with no knowledge of the implementation.  It is therefore _very_ important that your scenario read well.

hannahwhy commented on features/step_definitions/appointment_sheet_steps.rb:
```diff
@@ -0,0 +1,197 @@
+
+Given /^is a boy$/ do
+  @child_person.update_attribute(:sex_code, 1)
+end
```
This step isn't appointment-sheet specific.

Where possible, avoid grouping steps by scenario.  Strive to reuse existing steps, extending them if doing so is not too difficult.  (It's a judgment call; there is no hard-and-fast criteria.)  Building out specific steps for scenarios has a nasty habit of developing into scenarios that use slightly different steps to express the same actions and preconditions, which significantly impairs maintainability.

hannahwhy commented on features/step_definitions/appointment_sheet_steps.rb:
```diff
@@ -0,0 +1,197 @@
+  step "I should see #{name}"
+  step "I should see #{date}"
+  step "I should see #{event}"
+  step "I should see #{time}"
```
Use `steps`.

``` ruby
Then /^I should see ("[^"]*") scheduled for ("[^"]*") for a ("[^"]*") event at ("[^"]*")$/ do |name, date, event, time|
  steps %Q{
    Then I should see "#{name}"
    And I should see "#{date}"
 }
end
```

hannahwhy commented on features/step_definitions/appointment_sheet_steps.rb:
```diff
@@ -0,0 +1,197 @@
+  step "I should see #{time}"
+end
+
+Then /^I should see scheduled event ("[^"]*")$/ do |event|
```
Avoid `should`.  It's a noise word.

In the context of Cucumber (and RSpec), "Then I see" means the same thing as "Then I should see".  A violation of either expectation will be logged.

hannahwhy commented on features/support/fakeweb.rb:
```diff
@@ -8,7 +8,10 @@
+FakeWeb.register_uri(:get, /\/api\/v1\/subjects\/xfda-a78s-h83b\/schedules.json$/,
+                    :body => "#{Rails.root}/features/fixtures/fakeweb/build_activity_plan_response_for_appointment_sheet.json", :status => ["200", "OK"], :content_type => "application/json")
+
+#FakeWeb.register_uri(:get, /\/api\/v1\/subjects\/((?!registered_with_psc))/, :body => "Unknown", :status => ["401", "Unknown"])
```
If this isn't needed, eliminate it.

stgallagher committed:
deleted migration #3785

Appointment sheets aren't persisted.

stgallagher committed:
consolidated child attributes class #3785

stgallagher committed:
added birth_date and due_date to sheet #3785

stgallagher committed:
deleted unnecessary schema change #3785

stgallagher committed:
consolidated child attributes into a class #3785

stgallagher committed:
deleted self reflecting test

stgallagher committed:
forgot to take off the @javascript #3785

stgallagher commented:
Added due date and birth date to the appointment sheet per Sean's request. Fixed a few of the noted issues:
- deleted migration and schema changes
- deleted bad test from address spec

stgallagher committed:
deleted double "the" typo #3785

hannahwhy commented on app/models/appointment_sheet.rb:
```diff
@@ -0,0 +1,173 @@
+      return phone.phone_nbr
+    end
+    phone.phone_nbr.insert(-5, '-').insert(-9, '-') if phone
+  end
```
The formatting code here seems like it belongs on Telephone.  (It's formatting, right?  It looks like you're asking for a phone number in XXX-XXX-XXXX format.)

I suggest this:

``` ruby
phone.to_s
```

Or, if `#to_s` is already in use and can't be changed, something similar.  `#dashed` is one possibility.

hannahwhy commented on app/views/appointment_sheets/show.html.haml:
```diff
@@ -0,0 +1,132 @@
+              = "Birth date: #{@sheet.child_birth_dates[child_index]}"
+            - else
+              .not-found
+                Unknown birth_date
```
All of these if-elses can be handled with `||`, i.e.

``` ruby
"Birth date: #{birth_date}" || "Unknown birth date"
```

This is a good place to introduce helpers.

hannahwhy replied:
Also, it's "birth date", not "birth_date".

hannahwhy commented on app/views/appointment_sheets/show.html.haml:
```diff
@@ -0,0 +1,132 @@
+        - activities_for_event.each do |a|
+          .activity
+            %span{ :title => a.current_state}
+              &nbsp;
```
Express this in styles, not via non-breaking spaces.

hannahwhy commented on app/views/appointment_sheets/show.html.haml:
```diff
@@ -0,0 +1,132 @@
+              .not-found
+                No consents found for this child
+            - else
+              - @sheet.child_consents[child_index].uniq.each do |consent|
```
Why would you be getting duplicate consent records?

hannahwhy commented on app/views/appointment_sheets/show.html.haml:
```diff
@@ -0,0 +1,132 @@
+          No children found
+      - else
+        #children
+        - @sheet.children.each_index do |child_index|
```
Instead of this, you can avoid having to propagate `child_index` everywhere by building an intermediate object for each child and iterating over that.  That implies that each object in `@sheet.children` would contain all information for that child; you wouldn't have to dereference anything.

stgallagher committed:
extracted common css #3785

stgallagher committed:
using #titleize for address state print #3785

stgallagher committed:
refactored #participant_consent #3785

extracted the consent type determination into 
their own methods. Fixed a duplication bug.

stgallagher committed:
created Telephone#dashed #3785

To print phone numbers with dashes inserted.

stgallagher committed:
removed #uniq #3785

#uniq was covering up a bug. Fixed a typo. Using 
#titleize for state name.

stgallagher committed:
Removed 'should' from features #3785

And replaced successive "Then"s with "And"s.

stgallagher committed:
wrapped multi-step step definitions #3785

stgallagher committed:
placed non-specific steps in common directory #3785

stgallagher committed:
needed this step #3785

stgallagher commented:
Ready for review.

### Summary

Responding to 2nd round of feedback

##### Actions
- Removed `#uniq` from consent listing. This method was covering up a duplication bug that I addressed by changing `#participant_consents`
- Moved phone formatting logic to `Telephone` as `#dashed`
- Used `#titleize`for state string output
- Removed should from features
- Wrapped multi-step features
- Moved non-specific steps into `common_steps`
- Changed multiple successive 'Then's to 'And's
- Fixed `birth_date` typo 

##### Deferring
- The construction of an intermediate child object : this is actually a different (better) approach but involves a good deal of reworking my current approach, which I think works, just isn't as conceptually sound
- Extracting if-else in the view to helper methods: if I have time to return to this I can do it, but I don't think right now it's justified.

##### Remaining Issues
- Some of the Cucumber tests fail when I run `rake cucumber:all` locally 
- I need to determine if the FakeWeb route I commented out is used and where (it's signature was covering up a route I needed to fake out PSC#build_activity_plan

hannahwhy commented on app/controllers/appointment_sheets_controller.rb:
```diff
@@ -0,0 +1,28 @@
+  def event_time
+    event_activity = @participant_activity_plan.scheduled_activities.find do |sa|
+      sa.date == params[:date] &&
+      sa.event == @sheet.event_type.downcase.gsub(/ /,'_') &&
```
A regexp isn't necessary for this; `String#gsub` handles strings just fine.  You can also use `String#tr`:

``` ruby
sa.event == @sheet.event_type.downcase.tr(' ', '_')
```

stgallagher replied:
`#tr` reflects more specifically what I intended. Didn't know about that method. I've changed it in the next set of revisions.

hannahwhy commented on app/helpers/application_helper.rb:
```diff
@@ -135,7 +135,7 @@ def staff_list
 
   def build_staff_list
-    users = Aker.authority.find_users
+    users = Aker.configuration.authorities.find { |auth| NcsNavigator::Authorization::Core::Authority === auth }.find_users
```
Why not just instantiate an appropriate Authority for this purpose?

(I'd recommend moving this out of `ApplicationHelper`, too)

rsutphin replied:
> Why not just instantiate an appropriate Authority for this purpose?

Because if the authority requires configuration, this will get an instance which is known to be appropriately configured for the application and environment.

stgallagher committed:
remedying failing features #3785

A typo, a missing #id, and making a mock URI more
defined so it stops intercepting routes not 
intended for it

stgallagher commented:
Print sheet is getting better. If anyone knows how to specify "always scale to a single page" let me know please.
![Screen Shot 2013-04-18 at 4 10 26 PM](https://f.cloud.github.com/assets/801580/399268/b675d974-a86c-11e2-950c-fc439f675b51.png)

stgallagher committed:
print sheet css #3785

stgallagher committed:
using #tr instead of #gsub #3785

stgallagher committed:
added last contact comment #3785

stgallagher committed:
add contact comments css #3785

stgallagher committed:
methods need to handle nils 3785

stgallagher committed:
tweak the view #3785

stgallagher commented:
Ready for review.

### Summary

Added last contact comment. Formatted print view. Responded to 3rd round of feedback.

##### Actions
- Getting the last comment from all the contacts associated with the person and all the contacts associated with that person's events
- Added stylesheet to expressly handle printing the appointment sheet
- Used #tr instead of #gsub in AppointmentSheetController#event_time since it's more reflective of my intent
- Tweeked the view
- Inserting lines of contingency code to handle potential nil source objects

Final (for now) print style - (If anyone knows who to get rid of the annoying URL under "Next Event" I'd appreciate it)
![Screen Shot 2013-04-19 at 2 09 02 PM](https://f.cloud.github.com/assets/801580/403663/4ada5152-a927-11e2-843f-1e79421cc3d2.png)

stgallagher committed:
Extracting bug fix into it's own ticket #3785

This code is part of an unrelated bug fix. 
Fixing it in it's own ticket.

pfriedman commented:
Merged PR into `develop`
